libflux/flux-core/src/semantic/types.rs

```diff
@@ -82,15 +84,22 @@ impl PartialEq for PolyType {
 }
 
 impl Substitutable for PolyType {
-    fn apply_ref(&self, sub: &Substitution) -> Option<Self> {
+    fn apply_ref(&self, sub: &dyn Substituter) -> Option<Self> {
         // `vars` defines new distinct variables for `expr` so any substitutions applied on a
         // variable named the same must not be applied in `expr`
-        let sub = sub.without(&self.vars);
-        self.expr.apply_ref(&sub).map(|expr| PolyType {
-            vars: self.vars.clone(),
-            cons: self.cons.clone(),
-            expr,
-        })
+        self.expr
+            .apply_ref(&|var| {
+                if self.vars.contains(&var) {
+                    None
+                } else {
+                    sub.try_apply(var)
+                }
+            })
+            .map(|expr| PolyType {
+                vars: self.vars.clone(),
+                cons: self.cons.clone(),
+                expr,
+            })
```

Comment:
The way I understand this, is if the polytype had any quantified variables, we used to remove those variables from the substitution before applying it. 

Now we instead create a substituter that will filter out the quantified variables on the fly as we recurse into the type expression.

Do I have that right?

Reply:
Yep, it is the same effect in the end in that we create a "view" into the substitution that lacks those variables and therefore won't replace them